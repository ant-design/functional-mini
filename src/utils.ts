let logSwitch = true;

export const setLogSwitch = (on: boolean) => {
  logSwitch = on;
};

//@ts-expect-error
export const log = (...args) => {
  if (!logSwitch) return;

  console.log.apply(console, args);
};

//@ts-expect-error
export const error = (...args) => {
  console.error.apply(console, args);
};

export const time = (label?: string) => {
  if (!logSwitch) return;

  console.time(label);
};

export const timeEnd = (label?: string) => {
  if (!logSwitch) return;
  console.timeEnd(label);
};

//@ts-expect-error
export const shallowCompare = (obj1, obj2) => {
  if (obj1 === obj2) return true;
  if (typeof obj1 !== typeof obj2) return false;
  if (typeof obj1 !== 'object' || typeof obj2 !== 'object') return false;
  if (obj1 === null || obj2 === null) return false;
  if (Array.isArray(obj1) || Array.isArray(obj2)) return false;
  if (Object.keys(obj1).length !== Object.keys(obj2).length) return false;
  for (const key in obj1) {
    if (obj1[key] !== obj2[key]) return false;
  }
  return true;
};

// 这么做是为了测试方便，隔离一些全局变量
export enum EComponent2Status {
  VALID = 'VALID',
  INVALID = 'INVALID',
  UNKNOWN = 'UNKNOWN',
}

let component2Status = EComponent2Status.UNKNOWN;
export const getComponent2Status = () => {
  return component2Status;
};

export const updateComponent2Status = (status?: EComponent2Status) => {
  if (status) {
    component2Status = status;
    //@ts-expect-error
  } else if (typeof my !== 'undefined') {
    //@ts-expect-error
    component2Status = my.canIUse('component2')
      ? EComponent2Status.VALID
      : EComponent2Status.INVALID;
  }
};

let incrementalId = 0;
const instanceKeyId = '_functional_instance_id_';

//@ts-expect-error
export function getIdFromAppxInstance(appxInstance) {
  if (!appxInstance) throw new Error('appxInstance param is falsy');
  let id;
  // 支付宝端
  if (appxInstance.$id) {
    id = `${appxInstance.$id}-${appxInstance.$page?.$viewId || ''}-${
      appxInstance.is || 'page'
    }`; /* 页面 */
  } else if (appxInstance[instanceKeyId]) {
    id = appxInstance[instanceKeyId];
  } else {
    id = `instance-${incrementalId++}`;
    appxInstance[instanceKeyId] = id;
  }
  if (!id) throw new Error('failed to identify appxInstance.$id');
  return id;
}

export const instanceKeyPropNames = '_functional_instance_prop_names_';

export function mergeObjectKeys(
  ...objects: Record<string, unknown>[]
): string[] {
  let res: string[] = [];
  for (const iterator of objects) {
    res = res.concat(Object.keys(iterator));
  }
  return res;
}
