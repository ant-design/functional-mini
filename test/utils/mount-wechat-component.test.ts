import { expect, test } from 'vitest';
import { vi } from 'vitest';
import { mountWechatComponent } from './common';
import { delay } from '../utils/utils';

test('测试 mountWechatComponent', async () => {
  const c = vi.fn();
  const ob1 = vi.fn();
  const ob2 = vi.fn();
  const ob3 = vi.fn();
  const ob4 = vi.fn();
  const ob5 = vi.fn();

  const opt = {
    properties: {
      foo: {
        type: String,
        value: 'abc',
      },
      bar: {
        type: Number,
        value: null,
      },
      bbloon: {
        type: Boolean,
        value: false,
      },
    },
    data: {
      ddd: 'aaa',
    },
    lifetimes: {
      attached() {
        c();
        const self = this;
        setTimeout(() => {
          self.setData({ ddd: '2222' });
        }, 1);
      },
    },
    observers: {
      '**': ob1,
      'foo, bbloon': ob2,
      'bar, bbloon': ob3,
      not_exits: ob4,
      ddd: ob5,
    },
  };
  expect(c).toBeCalledTimes(0);
  const instance = mountWechatComponent(opt);
  expect(c).toBeCalledTimes(1);
  expect(instance.data).toEqual(instance.properties);
  expect(instance.data).toMatchSnapshot();

  // 各种 observers
  // 没变
  instance.updateProps({
    foo: 'abc',
    bar: 0,
    bbloon: false,
  });

  expect(ob1).toBeCalledTimes(0);
  expect(ob2).toBeCalledTimes(0);
  expect(ob3).toBeCalledTimes(0);
  expect(ob4).toBeCalledTimes(0);

  instance.updateProps({
    foo: 'abc',
    bar: 123,
    bbloon: false,
  });

  expect(ob1).toBeCalledTimes(1);
  expect(ob2).toBeCalledTimes(0);
  expect(ob3).toBeCalledTimes(1);
  expect(ob4).toBeCalledTimes(0);
  expect(instance.data).toMatchSnapshot();

  expect(ob5).toBeCalledTimes(0);
  await delay(5);
  expect(instance.data).toMatchSnapshot();
  expect(ob5).toBeCalledTimes(1);
});
