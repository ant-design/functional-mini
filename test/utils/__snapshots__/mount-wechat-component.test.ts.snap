// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`测试 mountWechatComponent 1`] = `
{
  "bar": 0,
  "bbloon": false,
  "ddd": "aaa",
  "foo": "abc",
}
`;

exports[`测试 mountWechatComponent 2`] = `
{
  "bar": 123,
  "bbloon": false,
  "ddd": "aaa",
  "foo": "abc",
}
`;

exports[`测试 mountWechatComponent 3`] = `
{
  "bar": 123,
  "bbloon": false,
  "ddd": "2222",
  "foo": "abc",
}
`;
